### PR TITLE
Add support for numeric font `weight`

### PR DESCRIPTION
**Pull request by archmoj**
This PR adds numeric `weight` options to various fonts.

Please note that `scattergl` traces has limited support of numeric values i.e. the values are mapped to either `bold` or `normal`.

TODO: 
 - [x] Opened https://github.com/openmaptiles/fonts/issues/24 considering current bug with extra bold using `Open Sans` with `mapbox`
 - [x] Add test and fall backs for text rendered into WebGL subplots e.g. 'mapbox'
 - [x] Add extra tests for `mapbox`
 - [x] Partial support in `scattergl` i.e. 400 --> 'normal', 700 --> 'bold' 
        Also opened https://github.com/jednano/css-font-weight-keywords/issues/1
 - [x] Support in `gl3d`
 - [x] Support `base64` `font.weight` in gl3d
 
@plotly/plotly_js 
cc: @ndrezn

### CODE REVIEW

File: src/lib/coerce.js [hunk 1]
```diff
@@ -114,8 +114,13 @@ exports.valObjectMeta = {
             'are coerced to the `dflt`.'
         ].join(' '),
         requiredOpts: [],
-        otherOpts: ['dflt', 'min', 'max', 'arrayOk'],
+        otherOpts: ['dflt', 'min', 'max', 'arrayOk', 'extras'],
```
Comment (emilykl) [inline src/lib/coerce.js:117]: @archmoj Is this adding a new functionality to `coerce`, where numeric (or other) properties can also be configured to accept a list of 'extra' values?

That's pretty cool -- is there anywhere we should document that internally?

Comment (archmoj) [reply src/lib/coerce.js:117]: Good question.
Yes that's true.
We have other types that support the `extras` option so I think it's clear internally.
But on plotly.py we need to see if this changes would be reflected properly by codegen.

Comment (emilykl) [reply src/lib/coerce.js:117]: Good point.

Makes me think, it would be awesome if we could build checking the Plotly.py codegen into the CI process. It would have to produce some kind of output that's easy to verify. It could run only if the `schema.json` has changed.

Comment (emilykl) [reply src/lib/coerce.js:117]: Perhaps add a line in the docstring here explaining `extras` for integers (like the explanation [here](https://github.com/plotly/plotly.js/blob/353efc4090727bb486120ad6ebca90679d673a3a/src/lib/coerce.js#L243) for `flaglist`)

File: test/image/baselines/zz-gl3d_font-weight-scatter.png
Comment (emilykl) [inline test/image/baselines/zz-gl3d_font-weight-scatter.png:1]: For this mock, maybe use a font that has more available weights, to test that they work properly in WebGL.

File: stackgl_modules/package.json [hunk 2]
```diff
@@ -22,7 +22,7 @@
     "gl-plot2d": "^1.5.0",
     "gl-plot3d": "^2.4.7",
     "gl-pointcloud2d": "^1.0.3",
-    "gl-scatter3d": "^1.3.0",
+    "gl-scatter3d": "github:gl-vis/gl-scatter3d#52e58adea4141c86b019f25d781725b54fb2f07c",
```
Comment (archmoj) [inline stackgl_modules/package.json:25]: https://github.com/gl-vis/gl-scatter3d/pull/22

File: src/lib/coerce.js (hunk 1 above)
Comment (archmoj) [reply src/lib/coerce.js:119]: For flag lists combination mean something. That's why it is commented.

**Comment by archmoj** [comment]
OK. This is ready for review.

File: test/image/baselines/zz-gl3d_font-weight-scatter.png
Comment (archmoj) [reply test/image/baselines/zz-gl3d_font-weight-scatter.png:1]: Actually I was interested to test it with the default font as that's what most users use at this moment.

File: src/traces/scattergl/convert.js [hunk 3]
```diff
@@ -190,7 +190,7 @@ function convertTextStyle(gd, trace) {
     if(
         isArrayOrTypedArray(tfs) ||
         Array.isArray(tff) ||
-        Array.isArray(tfw) ||
+        Lib.isArrayOrTypedArray(tfw) ||
```
Comment (emilykl) [inline src/traces/scattergl/convert.js:193]: is this different from `isArrayOrTypedArray` above?

Comment (archmoj) [reply src/traces/scattergl/convert.js]: They are identical. Good eye. Addressed in [82863fd](https://github.com/plotly/plotly.js/pull/6990/commits/82863fdf0cb4ccfdd610567670b34bc3751380c9).

File: src/lib/coerce.js [hunk 4]
```diff
@@ -100,6 +100,8 @@ exports.valObjectMeta = {
         requiredOpts: [],
         otherOpts: ['dflt', 'min', 'max', 'arrayOk'],
         coerceFunction: function(v, propOut, dflt, opts) {
+            if(isTypedArraySpec(v)) v = decodeTypedArraySpec(v);
```
Comment (archmoj) [inline src/lib/coerce.js:103]: Fix for handling `base64` encoded arrays. 
cc: #5230 and https://github.com/plotly/plotly.py/pull/4470

File: src/lib/coerce.js [hunk 5]
```diff
@@ -114,8 +116,15 @@ exports.valObjectMeta = {
...
         requiredOpts: [],
-        otherOpts: ['dflt', 'min', 'max', 'arrayOk'],
+        otherOpts: ['dflt', 'min', 'max', 'arrayOk', 'extras'],
         coerceFunction: function(v, propOut, dflt, opts) {
+            if((opts.extras || []).indexOf(v) !== -1) {
+                propOut.set(v);
+                return;
+            }
+
+            if(isTypedArraySpec(v)) v = decodeTypedArraySpec(v);
```
Comment (archmoj) [inline src/lib/coerce.js:126]: Fix for handling base64 encoded arrays.
cc: https://github.com/plotly/plotly.js/pull/5230 and https://github.com/plotly/plotly.py/pull/4470

File: src/lib/coerce.js [hunk 6]
```diff
@@ -156,6 +165,8 @@ exports.valObjectMeta = {
         requiredOpts: [],
         otherOpts: ['dflt', 'arrayOk'],
         coerceFunction: function(v, propOut, dflt) {
+            if(isTypedArraySpec(v)) v = decodeTypedArraySpec(v);
```
Comment (archmoj) [inline src/lib/coerce.js:168]: Fix for handling base64 encoded arrays.
cc: https://github.com/plotly/plotly.js/pull/5230 and https://github.com/plotly/plotly.py/pull/4470

File: src/lib/coerce.js [hunk 7]
```diff
@@ -198,6 +209,8 @@ exports.valObjectMeta = {
...
```
Comment (archmoj) [inline src/lib/coerce.js:212]: Fix for handling base64 encoded arrays.
cc: https://github.com/plotly/plotly.js/pull/5230 and https://github.com/plotly/plotly.py/pull/4470

File: src/traces/scattermapbox/constants.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+// Must use one of the following fonts as the family, else default to 'Open Sans Regular'
+// See https://github.com/openmaptiles/fonts/blob/gh-pages/fontstacks.json
+var supportedFonts = [
```
Comment (emilykl) [inline src/traces/scattermapbox/constants.js:5]: @archmoj Does Mapbox make their list of supported fonts public anywhere as part of their package? Just wondering if there's a way we can ensure this list is always up-to-date without updating it manually.

Comment (archmoj) [reply src/traces/scattermapbox/constants.js:5]: No in fact we use latest v1 version of `mapbox-gl`. Due to the licence changes of their v2+ no significant v1 is expected. On the other hand we may switch to `maplibre-gl-js` which is the continuation of `mapb0x-gl` v1. See #5578.
BTW we could possibly improve our docs for these fonts in a separate PR.

Comment (emilykl) [reply src/traces/scattermapbox/constants.js:5]: OK got it. Do you know to what extent `maplibre` supports different font weights? Basically, will adding these options make switching to `maplibre` significantly more complicated?

File: src/traces/scattermapbox/convert.js [hunk 9]
```diff
@@ -369,11 +370,58 @@ function arrayifyAttribute(values, step) {
...
+            else str += ' Thin';
+        } else if(parts[0] === 'Open' && parts[1] === 'Sans') {
+            str = 'Open Sans';
+            if(weight > 750) str += ' Extrabold';
+            else if(weight > 650) str += ' Bold';
+            else if(weight > 550) str += ' Semibold';
+            else if(weight > 350) str += ' Regular';
+            else str += ' Light';
+        } else { // Other families
+            str = 'Klokantech Noto Sans';
```
Comment (emilykl) [inline src/traces/scattermapbox/convert.js:407]: Doesn't this `else` mean the font will default to `Klokantech Noto Sans` if it's not `Metropolis` or `Open Sans`?

File: src/traces/scattermapbox/defaults.js [hunk 10]
```diff
@@ -8,44 +8,7 @@ var handleLineDefaults = require('../scatter/line_defaults');
...
-    'Open Sans Regular',
-    'Open Sans Semibold Italic',
-    'Open Sans Semibold',
-    'Klokantech Noto Sans Bold',
-    'Klokantech Noto Sans CJK Bold',
-    'Klokantech Noto Sans CJK Regular',
-    'Klokantech Noto Sans Italic',
-    'Klokantech Noto Sans Regular'
-];
+var isSupportedFont = require('./constants').isSupportedFont;
```
Comment (emilykl) [inline src/traces/scattermapbox/defaults.js:11]: Oh now I see, this list already existed but was just moved 😄 @archmoj

File: src/traces/scattermapbox/convert.js (hunk 9 above)
Comment (emilykl) [reply src/traces/scattermapbox/convert.js:407]: oh I see, maybe not, because it has already passed through the `supplyDefaults` step -- can you confirm @archmoj ?

File: src/traces/scattermapbox/defaults.js [hunk 11]
```diff
@@ -104,12 +67,14 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
...
     if(clusterEnabled || subTypes.hasText(traceOut)) {
+        var layoutFontFamily = layout.font.family;
+
         handleTextDefaults(traceIn, traceOut, layout, coerce,
             {
                 noSelect: true,
                 noFontVariant: true,
                 font: {
-                    family: supportedFonts.indexOf(layout.font.family) !== -1 ? layout.font.family : 'Open Sans Regular',
+                    family: isSupportedFont(layoutFontFamily) ? layoutFontFamily : 'Open Sans Regular',
```
Comment (emilykl) [inline src/traces/scattermapbox/defaults.js:77]: ok, naive question because I don't quite understand -- if the user passes just `Metropolis` (for example) as the font family, won't that register as "not supported" because the string `Metropolis` is not in the supported fonts list, and so it will get replaced with `Open Sans`?

File: src/traces/scattermapbox/convert.js (hunk 9 above)
Comment (archmoj) [reply src/traces/scattermapbox/convert.js:407]: Good point. No that shouldn't happen because further down we test and ensure the font is supported. 
I could restrict this `else` to make it read better?

Comment (archmoj) [reply src/traces/scattermapbox/convert.js:407]: Addressed in [bff00ac](https://github.com/plotly/plotly.js/pull/6990/commits/bff00ac8d60f0b154aaec75615621e63cb89c2f1).

File: src/traces/scattermapbox/defaults.js (hunk 11 above)
Comment (emilykl) [reply src/traces/scattermapbox/defaults.js:77]: Now that we have `weight` and `style` options, would it make sense to change that behavior? Since a user could pass `family: 'Metropolis', weight: 'bold'` and we could use that to determine they want the font face `Metropolis Bold`.

Comment (archmoj) [reply src/traces/scattermapbox/defaults.js:77]: Actually I was thinking about that and I'd be happy to work on it in a separate PR.
Obviously we don't want to break previous mocks. 
Instead we could work on "adding support for minimal family names in `mapbox`".

File: src/traces/scattergl/convert.js [hunk 12]
```diff
@@ -216,7 +216,7 @@ function convertTextStyle(gd, trace) {
         optsOut.font = {
             size: tfs * plotGlPixelRatio,
             family: tff,
-            weight: tfw,
+            weight: weightFallBack(tfw),
```
Comment (emilykl) [inline src/traces/scattergl/convert.js:219]: Why is this required if we are already disallowing numeric weight values at the schema level?

Comment (archmoj) [reply src/traces/scattergl/convert.js:219]: Good question. 
This is for `scattergl` with limited support of `weight` noting that for `arrayOk` attributes we do not check every element at the supply defaults step which could result in performance issues during interactions with the plot.
As a result this fall back is needed to map unsupported values in arrays to supported options.

File: src/traces/scattermapbox/convert.js [hunk 13]
```diff
@@ -369,11 +370,58 @@ function arrayifyAttribute(values, step) {
...
+            if(weight > 850) str += ' Black';
+            else if(weight > 750) str += ' Extra Bold';
+            else if(weight > 650) str += ' Bold';
+            else if(weight > 550) str += ' Semi Bold';
+            else if(weight > 450) str += ' Medium';
+            else if(weight > 350) str += ' Regular';
+            else if(weight > 250) str += ' Light';
+            else if(weight > 150) str += ' Extra Light';
+            else str += ' Thin';
+        } else if(parts.slice(0, 2).join(' ') === 'Open Sans') {
```
Comment (emilykl) [inline src/traces/scattermapbox/convert.js:399]: nit: Couldn't this be `family.startsWith('Open Sans')` ?

File: src/traces/scattergl/convert.js [hunk 14]
```diff
@@ -225,6 +225,12 @@ function convertTextStyle(gd, trace) {
     return optsOut;
 }
 
+function weightFallBack(w) {
+    if(w <= 1000) {
+        return w > 500 ? 'bold' : 'normal';
```
Comment (emilykl) [inline src/traces/scattergl/convert.js:230]: Shouldn't the threshold be 400?

File: test/image/baselines/zz-gl2d_font-weight-scatter.png
Comment (emilykl) [inline test/image/baselines/zz-gl2d_font-weight-scatter.png:1]: Is it not possible to use numeric weights in `gl2d`?

**Comment by emilykl** [comment]
@archmoj This is looking really good. 🎉 

Could you summarize which trace types have limited/no support for numeric font weights?

This still needs to be tested with the Plotly.py codegen, right?

File: test/image/baselines/zz-gl2d_font-weight-scatter.png
Comment (archmoj) [reply test/image/baselines/zz-gl2d_font-weight-scatter.png:1]: Yes. For `heatmapgl` gl2d scene the numeric values work nicely similar to `gl3d` scenes and `scatter3d`.
Also for `scattergl` (since it uses a different pipeline) it works using the fallback which maps the values to `bold` and `normal` i.e. limited support.

File: src/traces/scattergl/convert.js (hunk 12 above)
Comment (emilykl) [reply src/traces/scattergl/convert.js:219]: gotcha, thank you. Could you perhaps add that as a comment above the function definition?

File: src/traces/scattergl/convert.js (hunk 14 above)
Comment (archmoj) [reply src/traces/scattergl/convert.js:230]: 500 is `Medium`  looks very similar to `Regular`.
IMHO it's debatable between 450/500 to 550.

**Comment by archmoj** [comment]
> This still needs to be tested with the Plotly.py codegen, right?

Good call.
I could test that. In case of noticing a problem around integers we could switch to use `enumerated` values here.

**Comment by archmoj** [comment]
> @archmoj This is looking really good. 🎉
> 
> Could you summarize which trace types have limited/no support for numeric font weights?

Good call. Added to the PR description:
`scattergl` traces has limited support of numeric values i.e. the values are mapped to either `bold` or `normal`.

File: src/traces/scattermapbox/convert.js (hunk 13 above)
Comment (archmoj) [reply src/traces/scattermapbox/convert.js:399]: Sure. That may break IE11 which is fine as of today :)

File: src/traces/scattergl/convert.js (hunk 12 above)
Comment (archmoj) [reply src/traces/scattergl/convert.js:219]: Addressed in [48d057f](https://github.com/plotly/plotly.js/pull/6990/commits/48d057fad97215e6c3d3a9924741918a76cd84ee).

File: src/traces/scattergl/convert.js (hunk 14 above)
Comment (emilykl) [reply src/traces/scattergl/convert.js:232]: Right, ok, I guess if the only options are `normal` or `bold` we have to draw the line somewhere. Maybe just add a comment explaining why 500 is chosen as the threshold.

File: src/traces/scattermapbox/defaults.js (hunk 11 above)
Comment (emilykl) [reply src/traces/scattermapbox/defaults.js:77]: ok, that's fine to do in another PR. Would be nice to have!

**Review by emilykl — APPROVED** [review]

File: src/traces/scattermapbox/defaults.js (hunk 11 above)
Comment (archmoj) [reply src/traces/scattermapbox/defaults.js:77]: Tracked in #6993.

File: src/traces/scattermapbox/constants.js (hunk 8 above)
Comment (birkskyum) [reply src/traces/scattermapbox/constants.js:5]: @emilykl , there is a migration branch in here maplibre where we should be able to try out:
- #7015
